src/Network/MQTT/Client.hs

```diff
@@ -114,7 +114,8 @@ data MQTTClient = MQTTClient {
   , _connACKFlags :: TVar ConnACKFlags
   , _corr         :: TVar (Map BL.ByteString MessageCallback)
   , _cbM          :: MVar (IO ())
-  , _cbHandle     :: TMVar (Async ())
+  , _hasCbThread  :: TVar Bool
+  , _cbHandle     :: MVar (Async ())
```

Comment:
It might be possible to get away with just one variable containing the async handle, my last attempt wasn't safe though.

Reply:
I think we can move forward with this and I can take a pass at that.  It's not obvious why you'd need more than a `TVar (Maybe (Async ())` here, though it might be tedious to get all that in place.

Reply:
Should be pretty straightforward actually, I should've just done that. 
Let me know if you're busy, I don't mind changing it.

Reply:
You did some good work here I'm happy to tinker with it a bit.  :)

Testing is a bit hard here, *but* it's relatively easy in my mqttd project, so I might try to see if I can have a meaningful test over there before pushing, but in any case, I'll push a new version today.

Thanks a lot for doing this!